src/System.Linq/src/System/Linq/AppendPrepend.cs

```diff
@@ -0,0 +1,715 @@
+                }
+
+                return new Prepend1Iterator<TSource>(this, element);
+            }
+
+            public override int GetCount(bool onlyIfCheap)
+            {
+                IIListProvider<TSource> listProv = _source as IIListProvider<TSource>;
+                if (listProv != null)
+                {
+                    int count = listProv.GetCount(onlyIfCheap);
+                    return count == -1 ? -1 : count + _preCount + _appCount;
+                }
+
+                return !onlyIfCheap || _source is ICollection<TSource> || _source is ICollection ? _source.Count() + _preCount + _appCount : -1;
+            }
+
+            public override TSource[] ToArray()
+            {
+                int count = GetCount(onlyIfCheap: true);
+                if (count == -1)
+                {
+                    return EnumerableHelpers.ToArray(this);
+                }
+
+                TSource[] array = new TSource[count + _preCount + _appCount];
+                int index = 0;
+                for (int i = -_preCount; i != 0; ++i, ++index)
+                {
+                    array[index] = _prepended[i];
+                }
+
+                Debug.Assert(index == _preCount);
+                foreach (TSource item in _source)
+                {
+                    array[index] = item;
+                    ++index;
+                }
+
+                Debug.Assert(index == array.Length - _appCount);
+                for (int i = 0; i != _appCount; ++i, ++index)
+                {
+                    array[index] = _appended[i];
+                }
+
+                Debug.Assert(index == array.Length);
+                return array;
+            }
+
+            public override List<TSource> ToList()
+            {
+                int count = GetCount(onlyIfCheap: true);
+                List<TSource> list = count == -1 ? new List<TSource>() : new List<TSource>(count + _appCount + _preCount);
+                for (int i = -_preCount; i != 0; ++i)
+                {
+                    list.Add(_prepended[i]);
+                }
+
+                list.AddRange(_source);
+
+                for (int i = 0; i != _appCount; ++i)
+                {
+                    list.Add(_appended[i]);
+                }
+
+                return list;
+            }
+        }
+    }
+
+    internal class SharedCollection<T>
+    {
+        private const int MaxLength = 64;
+        private T[] _store;
+        private int _count;
+
+        public SharedCollection(T item)
+        {
+            _store = new[] { item };
+            _count = 1;
+        }
+
+        public SharedCollection(T first, T second)
+        {
+            _store = new[] { first, second };
+            _count = 2;
+        }
+
+        public T this[int index] => _store[index >= 0 ? index : index + _store.Length];
+
+        public bool Append(T item, int number)
+        {
+            if (number != _count | number >= MaxLength)
```

Comment:
Its an as-a-rule rather than a measured micro-opt. As in "as a rule" just comparing integers will be cheaper than the cost of branching to avoid doing so, but you have to measure to know for sure.